src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java

```diff
@@ -0,0 +1,86 @@
+ *
+ * This transform is intended for direct use in pipelines.
+ *
+ * The reference bases paired with each read can be customized by passing in a reference window function
+ * inside the {@link ReferenceDataflowSource} argument to {@link #add}. See
+ * {@link org.broadinstitute.hellbender.engine.dataflow.datasources.RefWindowFunctions} for examples.
+ */
+public class AddContextDataToReadSpark {
+    public static JavaPairRDD<GATKRead, ReadContextData> add(
+            final JavaRDD<GATKRead> reads, final ReferenceDataflowSource referenceDataflowSource,
+            final JavaRDD<Variant> variants) {
+        // Join Reads and Variants, Reads and ReferenceBases
+        JavaPairRDD<GATKRead, Iterable<Variant>> readiVariants = JoinReadsWithVariants.join(reads, variants);
+        JavaPairRDD<GATKRead, ReferenceBases> readRefBases = JoinReadsWithRefBases.addBases(referenceDataflowSource, reads);
+
+        // For testing we want to know that the reads from the KVs coming back from JoinReadsWithVariants.Join
+        // and JoinReadsWithRefBases.Pair are the same reads from "reads".
+        boolean assertsEnabled = false;
+        assert assertsEnabled = true; // Intentional side-effect!!!
+        // Now assertsEnabled is set to the correct value
+        if (assertsEnabled) {
+            assertSameReads(reads, readRefBases, readiVariants);
+        }
+
+        JavaPairRDD<GATKRead, Tuple2<Iterable<Iterable<Variant>>, Iterable<ReferenceBases>>> cogroup = readiVariants.cogroup(readRefBases);
+        return cogroup.mapToPair(in -> {
+            List<Iterable<Variant>> liVariants = Lists.newArrayList(in._2()._1());
+            List<Variant> lVariants = Lists.newArrayList();
+            if (!liVariants.isEmpty()) {
+                final Iterable<Variant> iVariant = Iterables.getOnlyElement(in._2()._1());
+                // It's possible for the iVariant to contain only a null variant, we don't
+                // want to add that to the ReadContextData.
+                final Variant next = iVariant.iterator().next();
+                if (next != null) {
+                    lVariants = Lists.newArrayList(iVariant);
+                }
+            }
+
+            ReferenceBases refBases = Iterables.getOnlyElement(in._2()._2());
+            ReadContextData readContextData = new ReadContextData(refBases, lVariants);
+            return new Tuple2<>(in._1(), readContextData);
+        });
+    }
+
+    private static void assertSameReads(final JavaRDD<GATKRead> reads,
+                                        final JavaPairRDD<GATKRead, ReferenceBases> readRefBases,
+                                        final JavaPairRDD<GATKRead, Iterable<Variant>> readiVariants) {
+        List<GATKRead> vReads = reads.collect();
+        Map<GATKRead, ReferenceBases> vReadRef = readRefBases.collectAsMap();
+        Map<GATKRead, Iterable<Variant>> vReadVariant = readiVariants.collectAsMap();
```

Comment:
These three lines assume that the distributed collections fit in memory, which won't always be the case. Can you drop the checks on these in-memory collections, since the other ones should be sufficient.